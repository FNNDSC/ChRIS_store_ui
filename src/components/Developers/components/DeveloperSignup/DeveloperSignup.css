#developer-signup-form {
    color: black;
}

#developer-signup-creating {
    vertical-align: super;
}

#developer-signup-creating {
    vertical-align: super;
}

.pf-c-card__body:hover {
    background-color: #ffffff;
}
