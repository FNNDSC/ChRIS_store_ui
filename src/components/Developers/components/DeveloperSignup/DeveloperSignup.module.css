.developer-signup-creating {
    vertical-align: super;
}
