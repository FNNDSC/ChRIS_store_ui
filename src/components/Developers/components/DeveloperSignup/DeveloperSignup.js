import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import {
  Form,
  FormGroup,
  ControlLabel,
  FormControl,
  HelpBlock,
  Button,
  Spinner,
} from 'patternfly-react';
import StoreClient from '@fnndsc/chrisstoreapi';
import { validate } from 'email-validator';
import './DeveloperSignup.css';
import ChrisStore from '../../../../store/ChrisStore';

/* inspired by http://bit.ly/2KycT4G */
const isTouchDevice = () => {
  if (('ontouchstart') in window || (window.DocumentTouch && document instanceof window.DocumentTouch)) {
    return true;
  }

  if (window.matchMedia) {
    const prefixes = ' -webkit- -moz- -o- -ms- '.split(' ');
    const query = ['(', prefixes.join('touch-enabled),('), 'heartz', ')'].join('');
    return window.matchMedia(query).matches;
  }

  return false;
};

export class DeveloperSignup extends Component {
  constructor(props) {
    super(props);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.state = {
      loading: false,
      error: {
        message: '',
        controls: '',
      },
    };
  }

  handleChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }

  handleSubmit(event) {
    const {
      username, email, password, passwordConfirm,
    } = this.state;
    const { store } = this.props;
    event.preventDefault();

    if (!username) {
      return this.setState({
        error: {
          message: 'Username is required',
          controls: ['username'],
        },
      });
    }

    if (!email || !validate(email)) {
      return this.setState({
        error: {
          message: 'A valid Email is required',
          controls: ['email'],
        },
      });
    }

    if (!password) {
      return this.setState({
        error: {
          message: 'Password is required',
          controls: ['password'],
        },
      });
    }

    if (!passwordConfirm) {
      return this.setState({
        error: {
          message: 'Confirmation is required',
          controls: ['confirmation'],
        },
      });
    }

    if (password !== passwordConfirm) {
      return this.setState({
        error: {
          message: 'Password and confirmation do not match',
          controls: ['password', 'confirmation'],
        },
      });
    }

    this.setState({
      loading: true,
      error: {
        message: '',
        controls: '',
      },
    }, () => store.set('userName')(username));

    return this.handleStoreLogin();
  }

  handleStoreLogin() {
    const { username, email, password } = this.state;
    const { store } = this.props;
    const storeURL = process.env.REACT_APP_STORE_URL;
    const usersURL = `${storeURL}users/`;
    const authURL = `${storeURL}auth-token/`;
    let authToken;

    return new Promise(async (resolve, reject) => {
      try {
        await StoreClient.createUser(usersURL, username, password, email);
      } catch (e) {
        /* TODO: JCC Enhance error handling */
        this.setState({
          loading: false,
          error: {
            message: 'This field must be unique',
            controls: ['username'],
          },
        });
        return resolve(e);
      }
      try {
        authToken = await StoreClient.getAuthToken(authURL, username, password);
      } catch (e) {
        return reject(e);
      }
      return this.setState({
        toDashboard: true,
      }, () => {
        store.set('authToken')(authToken);
        return resolve(authToken);
      });
    });
  }

  render() {
    const {
      error,
      loading,
      toDashboard,
    } = this.state;

    if (toDashboard) {
      return <Redirect to="/dashboard" />;
    }
    const disableControls = loading;
    return (
      <Form onSubmit={this.handleSubmit} noValidate>
        <p>{loading ? 'Creating' : 'Create'} a ChRIS Developer account:</p>
        <FormGroup
          controlId="username"
          validationState={error.controls.includes('username') ? 'error' : null}
        >
          <ControlLabel>
            Username
          </ControlLabel>
          <FormControl
            type="text"
            autoComplete="off"
            autoFocus={!isTouchDevice()}
            onChange={this.handleChange}
            name="username"
            disabled={disableControls}
          />
          <HelpBlock>
            { error.controls.includes('username') ? error.message : 'Enter your username' }
          </HelpBlock>
        </FormGroup>
        <FormGroup
          controlId="email"
          validationState={error.controls.includes('email') ? 'error' : null}
        >
          <ControlLabel>
            Email
          </ControlLabel>
          <FormControl
            type="email"
            autoComplete="off"
            onChange={this.handleChange}
            name="email"
            disabled={disableControls}
          />
          <HelpBlock>
            { error.controls.includes('email') ? error.message : 'Enter your email' }
          </HelpBlock>
        </FormGroup>
        <FormGroup
          controlId="password"
          validationState={error.controls.includes('password') ? 'error' : null}
        >
          <ControlLabel>
            Password
          </ControlLabel>
          <FormControl
            type="password"
            autoComplete="new-password"
            onChange={this.handleChange}
            name="password"
            disabled={disableControls}
          />
          <HelpBlock>
            { error.controls.includes('password') ? error.message : 'Enter your password' }
          </HelpBlock>
        </FormGroup>
        <FormGroup
          controlId="password-confirm"
          validationState={error.controls.includes('confirmation') ? 'error' : null}
        >
          <ControlLabel>
            Password Confirmation
          </ControlLabel>
          <FormControl
            type="password"
            autoComplete="new-password"
            onChange={this.handleChange}
            name="passwordConfirm"
            disabled={disableControls}
          />
          <HelpBlock>
            { error.controls.includes('confirmation') ? error.message : 'Confirm your password' }
          </HelpBlock>
        </FormGroup>
        <Spinner loading={loading} size="md" inline>
          {
            <Button bsStyle="primary" bsSize="large" type="submit" disabled={disableControls}>
            Create Account
            </Button>
          }
        </Spinner>{loading && <span className="developer-signup-creating">  Creating Account</span>}
      </Form>);
  }
}

export default ChrisStore.withStore(DeveloperSignup);

DeveloperSignup.propTypes = {
  store: PropTypes.objectOf(PropTypes.object).isRequired,
};
