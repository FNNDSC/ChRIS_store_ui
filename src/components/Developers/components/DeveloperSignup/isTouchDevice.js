import React from 'react';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import {
  Form,
  FormGroup,
  ControlLabel,
  FormControl,
  HelpBlock,
  Button,
  Spinner,
} from 'patternfly-react';
import _ from 'lodash';
import StoreClient from '@fnndsc/chrisstoreapi';
import { validate } from 'email-validator';
import './DeveloperSignup.css';
import './DeveloperSignup';
import ChrisStore from '../../../../store/ChrisStore';


const isTouchDevice = () => {
  if (('ontouchstart') in window || (window.DocumentTouch && document instanceof window.DocumentTouch)) {
    return true;
  }

  if (window.matchMedia) {
    const prefixes = ' -webkit- -moz- -o- -ms- '.split(' ');
    const query = ['(', prefixes.join('touch-enabled),('), 'heartz', ')'].join('');
    return window.matchMedia(query).matches;
  }

  return false;
};

export default isTouchDevice;