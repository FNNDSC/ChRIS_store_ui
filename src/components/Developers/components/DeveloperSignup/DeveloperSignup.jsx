import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import _ from 'lodash';
import { validate } from 'email-validator';
import { Form, Spinner } from '@patternfly/react-core';

import './DeveloperSignup.css';
import StoreClient from '@fnndsc/chrisstoreapi';
import Button from '../../../Button';
import ChrisStore from '../../../../store/ChrisStore';
import FormInput from '../../../FormInput';
import isTouchDevice from './isTouchDevice';

export class DeveloperSignup extends Component {
  constructor(props) {
    super(props);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
    const queryParams = new URLSearchParams(window.location.search);
    const emailVal = queryParams.get('email');
    this.state = {
      loading: false,
      error: {
        message: String(),
        controls: [],
      },
      username: String(),
      email: emailVal,
      password: String(),
      passwordConfirm: String(),
    };
  }

  handleChange(value, name) {
    this.setState({ [name]: value });
  }

  handleSubmit(event) {
    event.preventDefault()
    event.persist();
    const { username, email, password, passwordConfirm } = this.state;
    const { store } = this.props;

    if (!username) {
      return this.setState({
        error: {
          message: 'Username is required',
          controls: ['username'],
        },
      });
    }

    if (!email || !validate(email)) {
      return this.setState({
        error: {
          message: 'A valid Email is required',
          controls: ['email'],
        },
      });
    }

    if (!password) {
      return this.setState({
        error: {
          message: 'Password is required',
          controls: ['password'],
        },
      });
    }

    if (!passwordConfirm) {
      return this.setState({
        error: {
          message: 'Confirmation is required',
          controls: ['confirmation'],
        },
      });
    }

    if (password !== passwordConfirm) {
      return this.setState({
        error: {
          message: 'Password and confirmation do not match',
          controls: ['password', 'confirmation'],
        },
      });
    }

    this.setState(
      {
        loading: true,
        error: {
          message: '',
          controls: '',
        },
      },
      () => store.set('userName')(username)
    );

    return this.handleStoreLogin();
  }

  async handleStoreLogin() {
    const { username, email, password } = this.state;
    const { store } = this.props;
    const storeURL = process.env.REACT_APP_STORE_URL;
    const usersURL = `${storeURL}users/`;
    const authURL = `${storeURL}auth-token/`;
    let authToken;

    try {
      await StoreClient.createUser(usersURL, username, password, email);
    } catch (e) {
      if (_.has(e, 'response')) {
        if (_.has(e, 'response.data.username')) {
          this.setState({
            loading: false,
            error: {
              message: 'This username is already registered.',
              controls: ['username'],
            },
          });
        } else {
          this.setState({
            loading: false,
            error: {
              message: 'This email is already registered.',
              controls: ['email'],
            },
          });
        }
      } else {
        this.setState({
          loading: false,
        });
      }
      return e;
    }

    try {
      authToken = await StoreClient.getAuthToken(authURL, username, password);
    } catch (e) {
      return Promise.reject(e);
    }

    return this.setState(
      {
        toDashboard: true,
      },
      () => {
        store.set('authToken')(authToken);
        return authToken;
      }
    );
  }

  render() {
    const {
      error,
      loading,
      toDashboard,

      username,
      email,
      password,
      passwordConfirm,
    } = this.state;

    if (toDashboard) return <Redirect to="/dashboard" />;

    const disableControls = loading;
    return (
      <Form noValidate id="developer-signup-form" onSubmit={this.handleSubmit}>
        <FormInput
          formLabel="Username"
          fieldId="username"
          validationState={error.controls.includes('username') ? 'error' : 'default'}
          placeholder="Enter your username"
          inputType="text"
          id="username"
          fieldName="username"
          value={username}
          autofocus={!isTouchDevice}
          onChange={(val) => this.handleChange(val, 'username')}
          disableControls={disableControls}
          error={error}
        />
        <FormInput
          formLabel="Email"
          fieldId="email"
          validationState={error.controls.includes('email') ? 'error' : 'default'}
          placeholder="Enter your email"
          inputType="email"
          id="email"
          fieldName="email"
          value={email}
          onChange={(val) => this.handleChange(val, 'email')}
          disableControls={disableControls}
          error={error}
        />
        <FormInput
          formLabel="Password"
          fieldId="password"
          validationState={error.controls.includes('password') ? 'error' : 'default'}
          placeholder="Enter a password"
          inputType="password"
          id="password"
          fieldName="password"
          value={password}
          onChange={(val) => this.handleChange(val, 'password')}
          disableControls={disableControls}
          error={error}
        />
        <FormInput
          formLabel="Password Confirmation"
          fieldId="password-confirm"
          validationState={error.controls.includes('confirmation') ? 'error' : 'default'}
          placeholder="Re-type your password"
          inputType="password"
          id="password-confirm"
          fieldName="passwordConfirm"
          value={passwordConfirm}
          onChange={(val) => this.handleChange(val, 'passwordConfirm')}
          disableControls={disableControls}
          error={error}
        />
        <div style={{ padding: '1em 0' }}>
          {loading ? (
            <Spinner size="md" />
          ) : (
            <Button variant="primary" type="submit" loading={disableControls}>
              Create Account
            </Button>
          )}
          {loading && <span id="developer-signup-creating">Creating Account</span>}
        </div>
      </Form>
    );
  }
}

export default ChrisStore.withStore(DeveloperSignup);

DeveloperSignup.propTypes = {
  store: PropTypes.objectOf(PropTypes.object).isRequired,
};
