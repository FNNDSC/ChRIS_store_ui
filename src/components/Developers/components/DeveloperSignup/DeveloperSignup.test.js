import React from 'react';
import { shallow } from 'enzyme';
import DeveloperSignup from './DeveloperSignup';

describe('DeveloperSignup', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<DeveloperSignup />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render Form component', () => {
    expect(wrapper.find('Form')).toHaveLength(1);
  });

  it('should render 4 FormGroup components', () => {
    expect(wrapper.find('FormGroup')).toHaveLength(4);
  });

  it('should render 4 ControlLabel components', () => {
    expect(wrapper.find('ControlLabel')).toHaveLength(4);
  });

  it('should render 4 FormControl components', () => {
    expect(wrapper.find('FormControl')).toHaveLength(4);
  });

  it('should render 4 HelpBlock components', () => {
    expect(wrapper.find('HelpBlock')).toHaveLength(4);
  });
});
