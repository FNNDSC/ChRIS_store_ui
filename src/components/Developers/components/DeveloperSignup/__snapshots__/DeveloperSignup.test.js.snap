// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeveloperSignup should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
}
`;
